Allow to lookup using secret name and collection path

joaogbcravo opened the pull request:
Allows getting Bitwarden secrets using the name of the Bitwarden Secret and its location (in a collection) instead of an ID.

Lerentis commented:
Hi @joaogbcravo ,
thanks for the PR. I will merge #19 probably later today, so can you rebase your changes afterwards? Then i can take a look. In general this seems like a great addition. As i am not really using collections maybe you can answer my question if they can not be referenced by id already?

joaogbcravo commented:
Hey! Thanks for considering merging it.

Items in the collection can be picked using ID already. That's not the "problem". 

The problem is: it is not so user-friendly to grab the id from bitwarden user interface and commit it to a git repo. If you have a proper bitwarden organization structured by different collections, referencing them by them is much more user-friendly, as you can commit names to the git repo without knowing its ID à priori.

Lerentis reviewed (changes requested):
thanks for the explanation @joaogbcravo !
i understand where you are coming from. i chose the ID because it is unique and when i look at your PR you are already aware of the limitations that a reference by name would bring. I would ask you to document this (mentioning in the README should be sufficient here) and after a rebase we can go forward here. 

As the ID lookup is a one time action i still believe that reference by ID is the way to go in a declarative format as in k8s for a reliable functionality

Lerentis commented:
ah and with #19 , dont forget to bump the versions in the `Chart.yaml` and write a changelog entry :smile: 
i need to write a PR action for that :see_no_evil:

joaogbcravo commented:
I just rebased and added some more two "features":

1. Sync vault after unlock (otherwise it can not find new secrets (at least not when looking up by name))
2. Add the ownerreference on the metadata.

Lerentis commented:
hi @joaogbcravo ,

The periodic sync is something that i had in my mind as well. thanks for picking this up 😃 

i will have a more in depth look hopefully tomorrow. What i noticed already is definitely not a patch release anymore. i would suggest to bump the chart by a minor (should be major but as this is still a 0.x release minor should be fine) as it ships a new version of the CRD. The application as well as it ships basically a new feature, which you should mention in the README.

Lerentis commented:
terrible sorry about the delay. i completely missed your ping. having a look right now

Lerentis commented on charts/bitwarden-crd-operator/crds/bitwarden-secrets.yaml:
```diff
@@ -41,11 +41,20 @@ spec:
+              - required:
+                  - id
+              - required:
+                  - item_name
```
Should be camelCase as you wrote in the example

joaogbcravo replied:
> I guess this needs more testing (...)

I will recheck what cab be possible wrong

> as a last point i would put the periodic sync in a dedicated function  (...)

Yes please, do it. I will remove it from my MR.

Lerentis commented on charts/bitwarden-crd-operator/crds/bitwarden-secrets.yaml:
```diff
@@ -41,11 +41,20 @@ spec:
+                  - id
+              - required:
+                  - item_name
+                  - collection_path
```
this one as well camelCase please

Lerentis reviewed (changes requested):
I guess this needs more testing. i can not get it to work in the current state. already altered the CRD to match the example and tried to get a secret via collectionPath but it just failed with an invalid mac. as a side effect the lookup within fields scope seems to be broken in this as well. i am happy to give it another spin if you are done with testing this on your side.

as a last point i would put the periodic sync in a dedicated function and maybe add a dedicated cron type to it that is added on startup of the operator. if you are unsure how to do that i will pick this up as a separate task

joaogbcravo commented:
Closing because is replaced by https://github.com/Lerentis/bitwarden-crd-operator/pull/73
